### DISCUSSION TITLE
0-based vs 1-based pagination

### DISCUSSION THREAD

**Discussion by pvdbosch** [belgif/rest-guide#186] [General]
I got the question why the REST guide decided upon 1-based pagination (i.e. page 1 is the first one); can't seem to find a documented rationale.

From what I can find, following points could be made:

for zero-based:
* Spring's Pageable is zero-based by default. However, there seems to be a config to convert it to 1-based from REST calls:
  * https://stackoverflow.com/questions/39884860/how-to-configure-spring-boot-pagination-starting-from-page-1-not-0
* aligned with most programming languages

for one-based:
* same as what's displayed to the consumer

In other industry guides:
* National Bank of Belgium and Flanders:
  * recommend offset (= zero-based index of first item rather than page number) and limit query-parameter (so more similar to cursors)
    * https://github.com/NationalBankBelgium/REST-API-Design-Guide/wiki/Pagination-Rules-and-metadata
      * _update_:  "currentPage" (only in response) is [1-based](https://github.com/NationalBankBelgium/REST-API-Design-Guide/wiki/Pagination-Example)
    * https://vlaamseoverheid.atlassian.net/wiki/spaces/IKPubliek/pages/2076934790/Algemene+REST+API-specificaties
* DigiPolis Antwerpen: https://antwerp-api.digipolis.be/#/content/developers/paging?id=r-pq-002
  * one-based bc of consumer-friendliness

Does anyone know of any other arguments, existing guidelines, or framework compatibility?

**Comment by jpraet** [comment]
Changing the guideline at this point from 1-based to 0-based would be a compatibility nightmare I think.
So are we just looking for arguments to justify the initial decision to make it 1-based? :)

Consumer-friendliness is as good a reason as any I think. No mapping needed when displaying the pagination UI.
I see it semantically as the number of the current page, like in a book. It's not an offset. The page number of the first page is one.
